### PR TITLE
Add typescript-fetch client

### PR DESCRIPTION
**Pull request by pcvonz**
Screenshot 2024-04-11 at 10 33 47 AM

[Basically followed the steps in the conclusion section of the webpack doc on tree shaking](https://webpack.js.org/guides/tree-shaking/#conclusion). 

[This issue in the openapi-generator github](https://github.com/OpenAPITools/openapi-generator/issues/8881#issue-821415296) has some good info on how this works. 

If we migrate to the `typescript-fetch` client, the bundle size of our API code should be substantially smaller on the marketing site. The screenshot shows the size of using one endpoint (`listCustomerReviews`) on the marketing site. 

[feature/sc-1906](https://app.shortcut.com/off-the-record/story/1906/add-typescript-fetch-client)

### CODE REVIEW

**Review by JStein92 — APPROVED** [review]
Fantastic!

File: package.json [hunk 1]
```diff
@@ -12,13 +12,20 @@
...
     "generate:devo": "npm run generate:devo-angularjs",
     "generate:devo-angularjs-ci": "npm run download:devo && gulp codegen && java -jar /opt/openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i api/otr-service.json -o dist/typescript -g typescript-angularjs && tsc --project tsconfig.json && npm run clean-source",
     "generate:devo-angular-ci": "npm run download:devo && gulp codegen && java -jar /opt/openapi-generator-630/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i api/otr-service.json -o dist/angular -g typescript-angular",
     "generate:devo-angular": "npm run download:devo && gulp codegen &&  openapi-generator-cli version-manager set 6.3.0 && openapi-generator-cli generate -i api/otr-service.json -o dist/angular -g typescript-angular",
+    "generate:devo-typescript-fetch": "npm run download:devo && gulp codegen &&  openapi-generator-cli version-manager set 4.3.1 && openapi-generator-cli generate -i api/otr-service.json -o dist/typescript-fetch -g typescript-fetch --additional-properties=typescriptThreePlus=true && tsc --project tsconfig.fetch.json && npm run clean-source",
```
Comment (Morkos) [inline package.json:28]: we'll need to update the backend service CI pipeline to run this npm script instead.

File: tsconfig.fetch.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+      "moduleResolution": "node",
+      "module": "esnext",
```
Comment (Morkos) [inline tsconfig.fetch.json]: are there any compatibility issues with using a different module system we use on the consuming packages (e.g. marketing site)? Marketing site is still using commonJS and now we're using node for this typescript client package

**Review by Morkos — CHANGES_REQUESTED** [review]
Two comments and questions

File: package.json (hunk 1 above)
Comment (pcvonz) [reply package.json:28]: I was thinking that we'd have both options as we transition away from the angularJS client to the `typescript-fetch` client.

File: tsconfig.fetch.json (hunk 2 above)
Comment (pcvonz) [reply tsconfig.fetch.json]: The marketing site is actually using the same [module type](https://github.com/OffTheRecordApp/otr_marketing_site/blob/42d80fb5dae3627f31bd119831f1e070191ef889/tsconfig.json#L4). [Webpack will automatically make modules work in older browsers](https://webpack.js.org/guides/getting-started/#modules). The webpack documentation warns against using `commonjs` and requires the `ES2015` `import / export` syntax to enable tree shaking. 

`esnext` would probably be fine.. but for peace of mind I updated the module type to the lowest one that supports `dynamic imports` on the marketing site: https://github.com/OffTheRecordApp/otr_marketing_site/pull/399

File: package.json (hunk 1 above)
Comment (Morkos) [reply package.json:28]: I meant you'll need to add it to the CI pipeline so you can generate this client as well.

**Review by Morkos — CHANGES_REQUESTED** [review]
Three requests

1. This will be the third generated client in this package. Is anyone using the vanilla  angularJS javascript client? If not, let's remove it. I believe the ui -shared components or ui-shared services might be using it? If they are, then ignore this request but note which consumers are still using it

2. Any need to use the "*Raw" endpoints? If not, is there a way to configure to remove them so we don't have unnecessary code bloat

3. Can we inject the "domain" (devo/prod) backend URLs for all controllers to use just like how the previous generated clients do so?

**Comment by pcvonz** [comment]
> Three main requests
> 
>     1. This will be the third generated client in this package. Is anyone using the vanilla  angularJS javascript client? If not, let's remove it. I believe the ui -shared components or ui-shared services might be using it?

Not sure. Could we discuss removing it in a future PR?

>     2. Any need to use the "*Raw" endpoints? If not, is there a way to configure to remove them so we don't have unnecessary code bloat

Since we're using webpack for our front-end code this shouldn't increase the size bundle size. 

>     3. Can we inject the "domain" (devo/prod) backend URLs for all controllers to use just like how the previous generated clients do so?

Not sure why we'd do that here. The generated client includes a `Configuration` class we can use in the consumers:

```ts
  const conf = new Configuration({
      basePath: env.apiEndpoint // Whatever the endpoint should be
  });
  return new CourtControllerApi(conf);
```

**Comment by Morkos** [comment]
> > Three main requests
> > ```
> > 1. This will be the third generated client in this package. Is anyone using the vanilla  angularJS javascript client? If not, let's remove it. I believe the ui -shared components or ui-shared services might be using it?
> > ```
> 
> Not sure. Could we discuss removing it in a future PR?
> 
> > ```
> > 2. Any need to use the "*Raw" endpoints? If not, is there a way to configure to remove them so we don't have unnecessary code bloat
> > ```
> 
> Since we're using webpack for our front-end code this shouldn't increase the size bundle size.
> 
> > ```
> > 3. Can we inject the "domain" (devo/prod) backend URLs for all controllers to use just like how the previous generated clients do so?
> > ```
> 
> Not sure why we'd do that here. The generated client includes a `Configuration` class we can use in the consumers:
> 
> ```ts
>   const conf = new Configuration({
>       basePath: env.apiEndpoint // Whatever the endpoint should be
>   });
>   return new CourtControllerApi(conf);
> ```

Yep, you can create shortcut task to investigate and migrate consumers of the vanilla angularJS client. Add it to the backlog though

**Review by Morkos — APPROVED** [review]
Looks good. Remember to update the build spec in the CI pipeline

**Pull request merged**